### ISSUE TITLE
IOS - When using multiple lines for the Welcome message, after saving and returning, the app scrolls

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#30707]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.2.94.0
**Reproducible in staging?:**Y
**Reproducible in production?:**Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/30238
## Action Performed:
1. Log in with any acccount
2. Go to #announce or #admins chat
3. Click on header > Details > Settings > Welcome message
4. Use multiple lines for Welcome message
5. Save it
6. Go back to Welcome message again

## Expected Result:
When using multiple lines for the Welcome message, after saving and returning, the app should not scroll the message from the top to the bottom

## Actual Result:
When using multiple lines for the Welcome message, after saving and returning, the app scrolls the message from the top to the bottom

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/1b6d98ee-0eaa-498d-ae7d-3f7d512de1fe

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by ikevin127** [comment]
Other issues / discussions related to similar behaviour on iOS:
https://github.com/Expensify/App/pull/28790
https://github.com/Expensify/App/issues/20836

# Proposal

## Please re-state the problem that we are trying to solve in this issue

When using multiple lines for the Welcome message, after saving and returning back to editing, the app scrolls the multi-line input from the top to the bottom.

## What is the root cause of that problem?

On iOS, automatically focusing on a text input field when a page is loaded will, by default, scroll to the end of the input content. The `useFocusEffect` hook in the code below is setting the focus to the `welcomeMessageInputRef` field as soon as the ReportWelcomeMessagePage component is loaded.

https://github.com/Expensify/App/blob/3db5ff4a1924156c748291d25eb9d774d82f879e/src/pages/ReportWelcomeMessagePage.js#L65

This behavior occurs only on iOS and is not a bug but a feature. This is specifically designed to help the user start typing at the end of a multiline text without needing to place the cursor manually.

On top of this, we also have this piece of code in the same component:

https://github.com/Expensify/App/blob/3db5ff4a1924156c748291d25eb9d774d82f879e/src/pages/ReportWelcomeMessagePage.js#L109

Which does the following:
```
export default function updateMultilineInputRange(input, shouldAutoFocus = true) {
    if (!input) {
        return;
    }

    if (input.value && input.setSelectionRange) {
        const length = input.value.length;
        input.setSelectionRange(length, length);
        // eslint-disable-next-line no-param-reassign
        input.scrollTop = input.scrollHeight;
    }
}
```
And this is the actual fix for the problem because the function seen above sets the selection at the end of the multi-line text and then scrolls to it instead of calling `focus()`, but because of the `useFocusEffect` being called too which does call `focus()`, the scrolling is visible:

<details>
<summary>iOS: Native - Visible scrolling</summary>

https://github.com/Expensify/App/assets/56457735/6c6db96c-a6e8-4fdd-aa81-4a80953defc4

</details>

## What changes do you think we should make in order to solve the problem?

My proposal is to remove the `useFocusEffect` entirely which would fix the visible scrolling issue because it will be handled gracefully by the existing `updateMultilineInputRange(welcomeMessageInputRef.current)` code as it can be seen below.

 ### Videos

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/84134af1-8485-4f16-8889-65e94ba226a7

</details>

**Comment by ikevin127** [comment]
## Updated [proposal](https://github.com/Expensify/App/issues/30707#issuecomment-1789790417)
- changed in order to ensure the fix results in the same behaviour on all platforms

**Comment by isabelastisser** [comment]
Hey @cubuspl42, can you please review the proposal above? Thanks!

**Comment by cubuspl42** [comment]
@ikevin127 The `updateMultilineInputRange` is already platform-dependent (based on the `.js` / `.ios.js` file suffixes), so wrapping it with `platform === CONST.PLATFORM.` sounds redundant.

**Comment by ikevin127** [comment]
@cubuspl42 You're right, my bad. Second `updateMultilineInputRange` does not need to be wrapped, should be left untouched. Proposal updated.

**Comment by cubuspl42** [comment]
This is an iOS-specific issue.

> On top of this, we also have this piece of code in the same component:
> ...
> Which does the following:

Then, you provide the **non-iOS** implementation (from `index.js`) and draw conclusions.

Doesn't this invalidate your root cause analysis?

**Comment by ikevin127** [comment]
## Updated [proposal](https://github.com/Expensify/App/issues/30707#issuecomment-1789790417)
- updated root cause and fix

@cubuspl42 Yes, it does seem to invalidate my root cause analysis. Upon checking what's happening exactly, I think the root cause has to be updated.

The reason why this is happening on iOS: Native is because there's a race condition between `useFocusEffect` which calls `welcomeMessageInputRef.current.focus()` with a timeout of 300ms and the `updateMultilineInputRange(welcomeMessageInputRef.current)` which is called right after setting the `TextInput`'s ref, in the component return.

Why I'm saying this:
- when commenting out `updateMultilineInputRange(welcomeMessageInputRef.current)`, the bug described in this issue can be observed:

<details>
<summary>iOS: Native - Issue reproduced 100% rate</summary>

https://github.com/Expensify/App/assets/56457735/e81fd076-f4c8-4420-87b6-cc35dd1d62bf

</details>

which confirms the root cause of this issue: race condition between `welcomeMessageInputRef.current.focus()` called from `useFocusEffect` within a 300ms delay and `updateMultilineInputRange(welcomeMessageInputRef.current)` which is called after setting the `TextInput` ref.

(iOS platform-dependent version of `updateMultilineInputRange`)
**UpdateMultilineInputRange/index.ios.js**: 
```
export default function updateMultilineInputRange(input, shouldAutoFocus = true) {
    if (!input) {
        return;
    }

    if (shouldAutoFocus) {
        input.focus();
    }
}
```

this is the reason why I excluded `welcomeMessageInputRef.current.focus()` call from `useFocusEffect` for iOS because calling that `focus()` with a 300ms delay will, in some cases be called before the `updateMultilineInputRange(welcomeMessageInputRef.current)` and cause this issue.

This is how it looks on iOS: Native if we slow down the transition with my proposal's implementation:

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/a676d222-5065-40cd-a92a-bd8685fe3d8c

</details>

But then the reason I called `updateMultilineInputRange(welcomeMessageInputRef.current)` again in the `useFocusEffect` timeout is because on Firefox the `updateMultilineInputRange` called in return after the ref is set does not seem to work on Firefox:

<details>
<summary>MacOS: Firefox</summary>

https://github.com/Expensify/App/assets/56457735/4e259f70-bab1-4cc5-8cf7-ad1a7d300422

</details>

therefore my proposal's code changed would look like this:

```diff
    useFocusEffect(
        useCallback(() => {
            focusTimeoutRef.current = setTimeout(() => {
-               if (welcomeMessageInputRef.current) {
+               if (welcomeMessageInputRef.current && ![CONST.PLATFORM.IOS].includes(getPlatform())) {
                    welcomeMessageInputRef.current.focus();
+                   // called because on Firefox calling it when the input ref is set does not work
+                   updateMultilineInputRange(welcomeMessageInputRef.current);
                }
```

This ensures that on iOS: Native we fix the issue and we won't have a regression on Firefox.

**Comment by cubuspl42** [comment]
In the future, please don't add another comment with the proposal, as per the guidelines one should update the original comment, while adding this marker in a new comment:

```
## Proposal
[Updated](link to proposal)
```

Now, you can hide your previous comment as outdated.

![image](https://github.com/Expensify/App/assets/2590174/fce94cb4-5baf-487e-82d9-61acf099a574)

**Comment by cubuspl42** [comment]
```
[CONST.PLATFORM.IOS].includes(getPlatform()))
```

Is this a challenge to write `===` without using `===`? 🤔

I meant that, per the Expensify conventions (which I didn't invent), relying on `CONST.PLATFORM.*` constants is considered an anti-pattern. Please see for yourself how few usages there are in the codebase.

What is preferred, per the conventions, is relying on platform file suffixes `(.js` / `.ios.js` / etc.) and creating appropriate abstractions there.

> updateMultilineInputRange(welcomeMessageInputRef.current) which is called right after setting the TextInput's ref, in the component return.

I'm aware of this line, and I feel quite uncomfortable with it. I would like to see a solution that removes it (while keeping the problem it solves solved, of course...).

**Comment by ikevin127** [comment]
## Updated [proposal](https://github.com/Expensify/App/issues/30707#issuecomment-1789790417)
- changed code solution to use a lib with platform file suffixes and created appropriate abstractions there

@cubuspl42 I updated the proposal again, please let me know if this satisfies or if other changes are needed.

Unfortunately we cannot remove `updateMultilineInputRange(welcomeMessageInputRef.current)` which is called right after the TextInput's ref while maintaining the functionality and solving the problem because there's no better place to call this right after the ref was set then where it is right now, calling it in useFocusEffect or anywhere else with a timeout will break the functionality that we're trying to achieve.

**Comment by isabelastisser** [comment]
Hey @cubuspl42, can you please review the proposal above? Thanks!

**Comment by cubuspl42** [comment]
@ikevin127 

Thank you for the update!

While I appreciate the efforts and adjusting the solution to the feedback, I'm still not convinced by the solution. It feels extremely specific and work-aroundish.

I have a strong intuition that we miss the bigger picture. There's a race of two pieces of logic that _both_ affect focusing/scrolling/placing the cursor. I don't feel that adding `if`s with lengthy comments saying that "another component with timeout of 300s does this and that" is the way to go.

I'd like make the race disappear completely.

> when commenting out updateMultilineInputRange(welcomeMessageInputRef.current), the bug described in this issue can be observed [100% of the time]

Sure, but on the other hand, commenting out the `useFocusEffect` call makes the issue completely disappear **on iOS**, am I wrong? Could this observation lead us somewhere?

**Comment by ikevin127** [comment]
> Sure, but on the other hand, commenting out the useFocusEffect call makes the issue completely disappear on iOS, am I wrong? Could this observation lead us somewhere?

Yes, that's right but then nothing happens on Firefox and on web in general (any browser), we don't get the focus on the input. If say we add the `shouldAutoFocus` functionality back to the `index.js` like so:

```diff
-export default function updateMultilineInputRange(input, shouldAutoFocus = true) {
+export default function updateMultilineInputRange(input, shouldAutoFocus = false) {
    if (!input) {
        return;
    }

    if (input.value && input.setSelectionRange) {
        const length = input.value.length;
        input.setSelectionRange(length, length);
        // eslint-disable-next-line no-param-reassign
        input.scrollTop = input.scrollHeight;

+       if (!shouldAutoFocus) {
+           return;
+       }

+       input.focus();
    }
}
```

it still doesn't work on browsers.

Hopefully somebody is able to come with a proposal that satisfies all your requirements and also work as expected on all platforms. Unfortunately I cannot come up with a better one at this moment.

**Comment by cubuspl42** [comment]
> Yes, that's right but then nothing happens on Firefox and on web in general

Sure, but I didn't say that this is _the solution_. It's an observation.

My conclusion so far is that our cross-platform goal here is:

**Focus the input, as soon as possible, ensuring the cursor is at the end.**

We have various platform-dependent constraints here, and we might have to challenge the existing code.

I would expect a solid proposal that proves a full understanding of our focus-related problems and provides a clear, complex solution.

Good starting point:

- `useDelayedInputFocus` hook, which has no uses (?!)
- `useFocusEffect` hook usage across the project
  - One can perform a textual search for `useFocusEffect(` and put attention to the cases which integrate `setTimeout` + `CONST.ANIMATED_TRANSITION`. Some are **literally** copy-pasted. (around 10 cases total?) 
- `updateMultilineInputRange`
  - both the default (which should be `.web.js`, I think?) and the iOS variant
- an observation that on iOS the `CONST.ANIMATED_TRANSITION`-delayed focus is not needed and problematic, as we already focus early to work around another issue
- an observation that `useFocusEffect` in general means "call this thingy _every time_ this component is focused", while in many of our uses we _probably_ mean "we actually care about _the first_ time this component gets focuses, and only because we can't do what we want earlier."

@isabelastisser Would you raise the bounty? Although the problem might seem trivial, solving it correctly requires a bit of engineering, as I summed up above.

**Comment by isabelastisser** [comment]
@cubuspl42, it won't be possible to raise the bounty. This is the standard payment payment for all issues. Please let me know if we can move forward with the proposal above. Thanks!

**Comment by cubuspl42** [comment]
@isabelastisser Should the Melvin message above be updated? Since when is raising the bounty not an option?

No, I don't think we should move forward with the existing proposal.

**Comment by ikevin127** [comment]
## Updated [proposal](https://github.com/Expensify/App/issues/30707#issuecomment-1789790417)
- new solution w/ alternative after revising reviewer's [latest comment](https://github.com/Expensify/App/issues/30707#issuecomment-1809924843)

**Comment by isabelastisser** [comment]
Hey @cubuspl42, please review the updated proposal above. Thanks!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by cubuspl42** [comment]
Hey @isabelastisser, please answer [the question asked above](https://github.com/Expensify/App/issues/30707#issuecomment-1815398788). Thanks!

**Comment by Julesssss** [comment]
We can still discuss bounty increases, but typically this occurs once the engineer is assigned for final review of the proposal. I think we should wait until we have settled on a proposal before making this descision.

**Comment by cubuspl42** [comment]
@Julesssss

>  but typically this occurs

Not according to my experience with Expensify, it sounds like we're talking about two different projects. Let's continue this on Slack for now, as this is bigger than topic than this specific issue.

**Comment by Julesssss** [comment]
Okay, I thought the suggestion here was to increase the payout based on complexity.

So yes we should increase the bounty as we haven't yet received a satisfactory proposal

**Comment by situchan** [comment]
Another option is to float this around expert contributor group like Callstack.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by isabelastisser** [comment]
The bounty has been increased. Waiting for proposals!

**Comment by cubuspl42** [comment]
[I advertised the issue on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1701165571879629)

**Comment by cubuspl42** [comment]
🎀 👀 🎀

**Comment by cubuspl42** [comment]
> Decide whether any proposals currently meet our guidelines and can be approved as-is today

I don't think so

> If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities

> Current assignee @cubuspl42 is eligible for the Internal assigner

@techievivek I never entered this path, does it mean that I would be responsible for solving this issue, and you would review it?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by techievivek** [comment]
>  I never entered this path, does it mean that I would be responsible for solving this issue, and you would review the PR?

Aaah, no, that means that the internal engineer who is assigned to the GH should now start looking into the issue rather than keeping it external and hoping some contributor would work on it.

**Comment by techievivek** [comment]
It looks like this is no longer reproducible. @cubuspl42 are you still able to reproduce this issue?

**Comment by cubuspl42** [comment]
@techievivek Ouch. I checked and it's indeed not reproducible. 🙈 I should've caught up that earlier.
